placement/interface.go

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package placement
+
+// Algorithm places shards on hosts
+type Algorithm interface {
+	// InitPlacement initialize a sharding placement with RF = 1
+	BuildInitialPlacement(hosts []Host, ids []uint32) (Snapshot, error)
+
+	// AddReplica up the RF by 1 in the cluster
+	AddReplica(p Snapshot) (Snapshot, error)
+
+	// AddHost adds a host to the cluster
+	AddHost(p Snapshot, h Host) (Snapshot, error)
+
+	// RemoveHost removes a host from the cluster
+	RemoveHost(p Snapshot, h Host) (Snapshot, error)
+
+	// ReplaceHost replace a host with a new host
+	ReplaceHost(p Snapshot, leavingHost, addingHost Host) (Snapshot, error)
+}
+
+// DeploymentPlanner generates deployment steps for a placement
+type DeploymentPlanner interface {
+	// DeploymentSteps returns the deployment steps
+	DeploymentSteps(p Snapshot) [][]HostShards
+}
+
+// Snapshot describes how shards are placed on hosts
+type Snapshot interface {
+	// HostShards returns all HostShards in the placement
+	HostShards() []HostShards
+
+	// HostsLen returns the length of all hosts in the placement
+	HostsLen() int
+
+	// Replicas returns the replication factor in the placement
+	Replicas() int
+
+	// ShardsLen returns the number of shards in a replica
+	ShardsLen() int
+
+	// Shards returns all the unique shard ids for a replica
+	Shards() []uint32
+}
+
+// HostShards represents a host and its assigned shards
+type HostShards interface {
+	Host() Host
+	Shards() []uint32
+}
+
+// Host contains the information needed for placement
+type Host struct {
```

Comment:
hmm, there is really not much to be lazy-loaded for a Host, I kinda need the address and rack information always, I can make it an interface, but don't think there is much possibility or value for having a different struct/impl of this interface so I turned it into a struct.